### PR TITLE
Fixed-wing PIFF controller, different PID banks for FW/MC

### PR DESCRIPTION
**Pull request by digitalentity**
An experiment to make airplane PID less nervous.

As a part of experiment there is a new feed-forward component to PID control. D-term is removed. D gain now controls feed-forward component on airplanes (this is temporary to avoid invading too deep into the code)

If somebody wants to test it you need to change your PIDs as described:
`new P gain` = `old P gain` / 5-10
`I gain` - unchanged.
`new D gain` = `old P gain`

or in words:

1. Reduce P-gain significantly, 5-10 times, may even set it to zero
2. Don't change I-gain
3. Replace D-gain with your old P-gain

Go fly. Try to tune your airplane using mostly D-gain (it's a feed forward component that doesn't use gyroscope closed-loop control).

Also see https://github.com/iNavFlight/inav/issues/970

### CODE REVIEW

**Comment by DzikuVx** [comment]
Nice!

**Comment by oleost** [comment]
Exciting!

Is it possible to have this as an extra PID controller?

At least temporarily so it's easy to test, even release a final version with two controllers and then remove one if it's superior?

**Comment by digitalentity** [comment]
@oleost zero out D (feed-forward now) and you have an old PID controller (mostly).

D-term is almost useless on airplanes - it only creates servo jitter and doesn't do any good for stabilisations that's why I simply disabled D here.

**Comment by digitalentity** [comment]
As a last thing for today - just fixed a bug with TPA :smile:

**Comment by DzikuVx** [comment]
I'm happy to report, it works :)
But, results were not much better than just with "unilimited" Iterm. 
Cutting P by factor of 5-7 was too much, and flying wing was unstable in the air. At the end, I came up with FeedForward and P at almost the same level (P the same as before testing of this branch). But I was able to lower Igain and keep the same results as before.
My general feeling was that wing was more responsive, maybe even sligtly too responsive sometimes. 

Here are logs. File 6 is the best I came up with and File 1 are entry values (`new P` = `old P / 4` and `new D = old P`)

[wing_1_blackbox_log_2016-12-30_111010.TXT](https://github.com/iNavFlight/inav/files/678872/wing_1_blackbox_log_2016-12-30_111010.TXT)

[wing_2_blackbox_log_2016-12-30_111010.TXT](https://github.com/iNavFlight/inav/files/678873/wing_2_blackbox_log_2016-12-30_111010.TXT)

[wing_3_blackbox_log_2016-12-30_111010.TXT](https://github.com/iNavFlight/inav/files/678874/wing_3_blackbox_log_2016-12-30_111010.TXT)

[wing_4_blackbox_log_2016-12-30_111010.TXT](https://github.com/iNavFlight/inav/files/678876/wing_4_blackbox_log_2016-12-30_111010.TXT)

[wing_5_blackbox_log_2016-12-30_111010.TXT](https://github.com/iNavFlight/inav/files/678877/wing_5_blackbox_log_2016-12-30_111010.TXT)

[wing_6_blackbox_log_2016-12-30_111010.TXT](https://github.com/iNavFlight/inav/files/678878/wing_6_blackbox_log_2016-12-30_111010.TXT)

Perhaps it would behave different on heavier and more stable airplane

**Comment by DzikuVx** [comment]
@digitalentity What do we do with this? I'd say it's safe to fly. @oleost any chance for testing this? How is weather?

**Comment by oleost** [comment]
Got it loaded on my airplane ready to test for awhile now. But the weather doesn't permit any test flying. And now I'm in work period for a week aswell.

**Comment by digitalentity** [comment]
@DzikuVx @oleost we need to decide whether we want to change meaning of D for airplanes or change the structure of PID config to include new PIDs for fixed-wing aircrafts. I'm in favor of the second option.

Initally MSP can stay unchanged and substitude PID banks for MR and FW based on mixer type. What do you think?

**Comment by DzikuVx** [comment]
Yeap, PID part is quite flexible in terms of MSP so as far as I'm concerned, we can go that way

**Comment by digitalentity** [comment]
Currently we have:
```
    PIDROLL,
    PIDPITCH,
    PIDYAW,
    PIDALT,
    PIDPOS,
    PIDPOSR,
    PIDNAVR,
    PIDLEVEL,
    PIDMAG,
    PIDVEL,
```

I'm going to rename them for better clarity:

| Old name | New name   | Description                         |
|----------|------------|-------------------------------------|
| PIDROLL  | PIDROLL    | roll angular rate                   |
| PIDPITCH | PIDPITCH   | pitch angular rate                  |
| PIDYAW   | PIDYAW     | yaw angular rate                    |
| PIDALT   | PIDZ       | Z-coordinate (altitude) controller  |
| PIDPOS   | PIDXY      | XY-coordinate (position) controller / will replace NavR for airplanes |
| PIDPOSR  | PIDXYRATE  | XY-rate (velocity) controller / unused on airplanes      |
| PIDNAVR  | removed    |                                     |
| PIDLEVEL | PIDLEVEL   | Self-leveling controller            |
| PIDMAG   | PIDHEADING | Heading controller                  |
| PIDVEL   | PIDZRATE   | Z-rate (climb rate) controller      |

I'll also make two sets of these PIDs in one PID profile - this will allow us to have completely separate PID settings for MR and FW aircrafts.

**Comment by oleost** [comment]
I like the idea to separate MC and airplanes PIDs. Also with VTOL in mind this will be great addition.

**Comment by digitalentity** [comment]
@martinbudden can you have a look at 52d4b88 regarding code correctness?

**Comment by digitalentity** [comment]
We now have two banks of PIDs per each profile. 

Active bank is selected according to `STATE(FIXED_WING)`. CLI names for PIDs have also changed - both banks are accessible via CLI, but only active bank is accessible via MSP `MSP_PID` and `MSP_SET_PID` API.

When we'll change the Configurator we need to keep backward compatibility - NavR PID is no more in this PR (renamed to PIDSURFACE and not used). Basically PID tab will have to behave differently in 1.5 and 1.6

**Comment by digitalentity** [comment]
Bench-tested, appears to be functioning properly

**Comment by martinbudden** [comment]
If we adding a new bank of PIDs, then I think we should correct the PID arrays, which I've always thought are the "wrong way around". Ie instead of:
```
typedef struct pidProfile_s {
    uint8_t P8[PID_ITEM_COUNT];
    uint8_t I8[PID_ITEM_COUNT];
    uint8_t D8[PID_ITEM_COUNT];
...
```
we should have
```
typedef struct pid8_s {
    uint8_t P;
    uint8_t I;
    uint8_t D;
} pid8_t;

typedef struct pidProfile_s {
    pid8_t pid8[PID_ITEM_COUNT];
...
```
There is no logical reason to group all the "Ps" together in an array.

Although slightly disruptive, this will make it easier to pass PIDs around, initialise them and also will make the code read easier. I think it will probably also make handling the PID banks easier.

**Comment by digitalentity** [comment]
@martinbudden I agree - will change since we are already touching the PIDs

**Comment by digitalentity** [comment]
@martinbudden accessing an element of `pidBank()` looks a bit odd: `pidBank()[axis].P`. Any suggestion how to structure it?

**Comment by digitalentity** [comment]
@martinbudden please have a look

**Comment by digitalentity** [comment]
Regular expressions are cool :smile:

File: src/main/flight/pid.h [hunk 1]
```diff
@@ -76,13 +83,16 @@ typedef struct pidProfile_s {
 
 PG_DECLARE_PROFILE(pidProfile_t, pidProfile);
 
+static inline const pid8_t * pidBank() { return STATE(FIXED_WING) ? pidProfile()->bank_fw : pidProfile()->bank_mc; }
```
Comment (martinbudden) [inline src/main/flight/pid.h]: You could define this as something like:
```
static inline const pid8_t * pidBank(int i) { return STATE(FIXED_WING) ? pidProfile()->bank_fw[i] : pidProfile()->bank_mc[i]; }
```
then you could use (eg):
```
pidBank(PID_ROLL)->I
```
Not 100% sure the syntax is right (haven't run it through the compiler), but you get the idea.
            
  Write

Comment (digitalentity) [reply src/main/flight/pid.h]: @martinbudden I think `pidBank(PID_ROLL)` is somewhat unintuitive. I would prefer `pidBank` to return a structure. I believe some other parameters might be part of `PID bank` as well, i.e. `dterm_setpoint_weight` or `yaw_p_limit`.

**Comment by digitalentity** [comment]
Ok, this no longer builds on `CRAZEPONYMINI` target. Can anybody remind me why we need this target at all?

**Comment by DzikuVx** [comment]
@digitalentity we do not have a single event from this target in configurator for last 3 months. I'd say: drop it

**Comment by digitalentity** [comment]
Looks like everything is functioning now. Merging

**Comment by oleost** [comment]
https://github.com/iNavFlight/inav/wiki/Hardware-and-feature-support-map

Needs to be updated, I'm on vacation now so only on mobile phone so rather not do it.

**Comment by digitalentity** [comment]
None of the removed features are listed in that table so nothing to update.

**Comment by oleost** [comment]
Yes I saw that. But we need it written down somewhere?

24. jan. 2017 11.34 a.m. skrev "Konstantin Sharlaimov" <
notifications@github.com>:

**Comment by digitalentity** [comment]
I'll handle that.

File: src/main/flight/pid.c [hunk 2]
```diff
@@ -263,74 +337,40 @@ void updatePIDCoefficients(const pidProfile_t *pidProfile, const controlRateConf
...
-        pidState[axis].kI = pidProfile->I8[axis] / FP_PID_RATE_I_MULTIPLIER;
-        pidState[axis].kD = pidProfile->D8[axis] / FP_PID_RATE_D_MULTIPLIER;
-
-        // Apply TPA
         if (STATE(FIXED_WING)) {
             // Airplanes - scale all PIDs according to TPA
-            pidState[axis].kP *= tpaFactor;
-            pidState[axis].kI *= tpaFactor;
-            pidState[axis].kD *= tpaFactor * tpaFactor;     // acceleration scales with speed^2
+            pidState[axis].kP  = pidBank()->pid[axis].P / FP_PID_RATE_P_MULTIPLIER  * tpaFactor * tpaFactor;
```
Comment (lkaino) [inline src/main/flight/pid.c:347]: Why is TPA applied twice to P term? Could add a comment.

Comment (digitalentity) [reply src/main/flight/pid.c:347]: This logic is borrowed from APM and PX4 - they both apply scaling squared to error-driven P-term and plain scaling to feed-forward. We'll see if this logic should be changed - we don't have much feedback of PIFF yet.

**Pull request merged**